DynamoPathManager complete removal

Benglin opened the pull request:
Background
-----
`DynamoPathManager` is now completely removed (like, hooray).

Changes
-----
1. Parameter-less `Load` and `Save` methods on `PreferenceSettings` class assumes the XML file to be stored in certain location. The behavior was dependent on settings file path in `DynamoPathManager`, but since it is managed outside of `PreferenceSettings`, its outcome is in-deterministic. This pull request enforces the path to be specified by the owner of `PreferenceSettings` object.

2. I choose to keep `DocumentationServices`, `LibraryCustomizationServices`, `XmlDocumentationExtensions` as `static` classes for now, otherwise that will bring the rest of the world into this PR. With that decision made, `IPathManager` needs to be passed into the them to help with path resolution.

3. None of the callers of `TypedParameter` constructor pass in `FunctionDescriptor`, they are always `null`. This pull request removes that redundant constructor. A new method `TypedParameter.UpdateFunctionDescriptor` is defined so that `FunctionDescriptor` can be updated after a `TypedParameter` object is constructed. During the call to `UpdateFunctionDescriptor` call, an optional `IPathManager` can be specified to help locating documentation assembly for the given `TypedParameter`.

4. `DynamoModel` now initializes and owns an instance of `PathManager` object.

5. `DynamoModel.StartConfiguration.DynamoCorePath` is retained, but mostly not used by most clients except Revit tests. The new class `RemoteTestSessionConfig` that @ikeough introduced requires a way to specify an alternative folder where `DynamoCore.dll` can be located. For most clients that do not specify this value, the default core path would be the folder that contains `DynamoCore.dll` that is executing the code.

Notes to Reviewer
-----
Hi @pboyer, I have no idea how to make this any smaller now that we have removed `DynamoPathManager`, the kind of all-or-nothing-at-all operation. Nearly half of the files are in tests, and I have ran all tests locally to make sure they remain passing. So perhaps you may want to focus only on actual code changes. To make reviewing even easier, I'm documenting the changes along the way, if existing comments are not already clear.

Your allocated time for review now comes in handy. Again, thanks for the additional pair of eyes! :)

Benglin commented on src/DynamoCore/Core/DynamoLoader.cs:
```diff
@@ -65,10 +65,12 @@ private void OnAssemblyLoaded(Assembly assem)
         /// the bin/nodes directory. Add the types to the searchviewmodel and
```
Additional `nodes` directories are now being passed to `DynamoLoader` instead of it having to depend on some `static` list.

Benglin commented on src/DynamoCore/Core/PreferenceSettings.cs:
```diff
@@ -21,7 +21,6 @@ namespace Dynamo
     public class PreferenceSettings : NotificationObject, IPreferences
```
Parameter-less `Load` and `Save` methods on `PreferenceSettings` class assumes the XML file to be stored in certain location. The behavior was dependent on settings file path in `DynamoPathManager`, but since it is managed outside of `PreferenceSettings`, its outcome is in-deterministic. This change enforces the path is specified by the owner of `PreferenceSettings` object.

Benglin commented on src/DynamoCore/Library/DocumentationServices.cs:
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
```
I choose to keep `DocumentationServices` a `static` class for now, otherwise that will bring the rest of the world into this PR. With that decision made, `IPathManager` needs to be passed into the methods to help with path resolution.

Benglin commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -3,7 +3,7 @@
 using System.IO;
```
Same decision here, `LibraryCustomizationServices` is kept as `static` for now, so `IPathManager` needs to be passed into them for now.

Benglin commented on src/DynamoCore/Library/LibraryCustomization.cs:
```diff
@@ -4,6 +4,7 @@
 using System.Reflection;
```
Same decision here, `LibraryCustomizationServices` is kept as `static` for now, so `IPathManager` needs to be passed into them for now.

Benglin commented on src/DynamoCore/Library/LibraryServices.cs:
```diff
@@ -34,16 +34,18 @@ public class LibraryServices : LogSourceBase, IDisposable
 
```
`LibraryServices` is keeping a reference to `IPathManager` to help with resolving library paths.

Benglin commented on src/DynamoCore/Library/TypedParameter.cs:
```diff
@@ -1,42 +1,33 @@
 ﻿using System;
```
None of the callers of `TypedParameter` constructor pass in `FunctionDescriptor`, they are always `null`. This pull request removes that redundant constructor. A new method `TypedParameter.UpdateFunctionDescriptor` is defined so that `FunctionDescriptor` can be updated after a `TypedParameter` object is constructed. During the call to `UpdateFunctionDescriptor` call, an optional `IPathManager` can be specified to help locating documentation assembly for the given `TypedParameter`.

Benglin commented on src/DynamoCore/Library/XmlDocumentationExtensions.cs:
```diff
@@ -1,10 +1,11 @@
 ﻿using System;
```
`XmlDocumentationExtensions` remains a `static` class, so `IPathManager` has to be passed in.

Benglin commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -361,8 +367,8 @@ protected virtual void PostShutdownCore(bool shutdownHost)
             public string Context { get; set; }
-            public string DynamoCorePath { get; set; }
             public IPreferences Preferences { get; set; }
+            public IPathResolver PathResolver { get; set; }
```
`DynamoModel.StartConfiguration.DynamoCorePath` is removed, and replaced by `IPathResolver` implementation from clients. Internally, `DynamoModel` will always retain this path as the containing folder of `DynamoCore.dll` for all purposes of path resolution.

Benglin commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -390,14 +396,6 @@ public static DynamoModel Start(StartConfiguration configuration)
-            }
-
-            if (configuration.Preferences == null)
-                configuration.Preferences = new PreferenceSettings();
```
See `CreateOrLoadPreferences` below for clarity.

Benglin commented on src/DynamoCore/PackageManager/PackageLoader.cs:
```diff
@@ -17,6 +17,7 @@ public struct LoadPackageParams
         public string Context { get; set; }
```
`PackageLoader` would need an instance of `IPathManager` to help with its path resolution.

Benglin commented on src/DynamoCoreWpf/Services/UsageReportingManager.cs:
```diff
@@ -54,17 +54,18 @@ private set
             {
```
Usage reporting now makes use of the per-instance preference settings path.

Benglin commented on src/DynamoCoreWpf/TestInfrastructure/CustomNodeCompatibilityMutator.cs:
```diff
@@ -236,14 +236,16 @@ public List<Type> LoadAllTypesFromDynamoAssemblies()
 
```
`DynamoModel.PathManager` now stores and manages the list of `nodes` folders internally.

Benglin commented on src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:
```diff
@@ -6,6 +6,7 @@
 using System.Text.RegularExpressions;
```
`IconServices` needs `IPathManager` to help locating icon resource files.

Benglin commented on src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:
```diff
@@ -791,11 +791,12 @@ private void TextBoxBase_OnTextChanged(object sender, TextChangedEventArgs e)
         /// </summary>
```
`DynamoModel.PathManager.SamplesDirectory` now points to the per-instance sample folder.

Benglin commented on src/DynamoSandbox/Program.cs:
```diff
@@ -53,18 +54,19 @@ internal class Program
-                    GeometryFactoryPath = geometryFactoryPath,
-                    Preferences = PreferenceSettings.Load()
+                    PathResolver = pathResolver,
+                    GeometryFactoryPath = geometryFactoryPath
```
`PreferenceSettings` is loaded within `DynamoModel` if none is specified.

Benglin commented on src/DynamoUtilities/AssemblyHelper.cs:
```diff
@@ -1,12 +1,24 @@
 ﻿using System;
```
`AssemblyHelper` is used for test cases. Each test fixture `Setup` now creates an instance of this class, registering to `AppDomain.CurrentDomain.AssemblyResolve` event at creation time, and unregistering from this event once `TearDown` happens.

Benglin commented on src/DynamoUtilities/DynamoPathManager.cs:
```diff
@@ -1,277 +0,0 @@
-﻿using System;
```
Some classes are better off removed ;)

pboyer replied:
That must have been satisfying. :)

Benglin commented on test/DynamoCoreTests/DynamoViewModelUnitTest.cs:
```diff
@@ -94,9 +94,7 @@ protected void VerifyModelExistence(Dictionary<string, bool> modelExistenceMap)
         protected void StartDynamo()
```
From this point onwards it's all test cases.

pboyer commented on src/DynamoCore/Core/PreferenceSettings.cs:
```diff
@@ -167,9 +166,12 @@ public bool Save(string filePath)
         /// </summary>
         /// <returns>Whether file is saved or error occurred.</returns>
-        public bool Save()
+        public bool SaveInternal(string preferenceFilePath)
```
I think this parameter deserves an argument description.

pboyer commented:
Hi @Benglin!  Nicely done - this wasn't that bad to review at all.  I only have some very minor comments.  LGTM!

Benglin commented:
Thanks again, @pboyer. I have addressed your comments about `bool SaveInternal` above, except that if the path is a non-existence file, the file will be created, so I didn't quite check to see if the path represents an existing file path.

pboyer commented:
OK - sorry if I overlooked something.
